src/backend/distributed/transaction/distributed_transaction_management.c

```diff
@@ -0,0 +1,483 @@
+
+	/* we'll use mutex while accessing the backend shmem */
+	SpinLockInit(&MyDistributedTransactionBackend->mutex);
+
+	LWLockRelease(&distributedTransactionShmemData->lock);
+}
+
+
+/*
+ * UnSetDistributedTransactionId simply acquires the mutex and resets
+ * the backend's distributed transaction data in shared memory to the
+ * initial values.
+ */
+void
+UnSetDistributedTransactionId(void)
+{
+	if (MyDistributedTransactionBackend)
+	{
+		SpinLockAcquire(&MyDistributedTransactionBackend->mutex);
+
+		MyDistributedTransactionBackend->databaseId = 0;
+		MyDistributedTransactionBackend->transactionId.initiatorNodeIdentifier = 0;
+		MyDistributedTransactionBackend->transactionId.transactionId = 0;
+		MyDistributedTransactionBackend->transactionId.timestamp = 0;
+
+		SpinLockRelease(&MyDistributedTransactionBackend->mutex);
+	}
+}
+
+
+/*
+ * GetCurrentDistributedTransctionId reads the backend's distributed transaction id and
+ * returns a copy of it.
+ */
+DistributedTransactionId *
+GetCurrentDistributedTransctionId(void)
+{
+	DistributedTransactionId *currentDistributedTransactionId =
+		(DistributedTransactionId *) palloc(sizeof(DistributedTransactionId));
+
+	SpinLockAcquire(&MyDistributedTransactionBackend->mutex);
+
+	currentDistributedTransactionId->initiatorNodeIdentifier =
+		MyDistributedTransactionBackend->transactionId.initiatorNodeIdentifier;
+	currentDistributedTransactionId->transactionId =
+		MyDistributedTransactionBackend->transactionId.transactionId;
+	currentDistributedTransactionId->timestamp =
+		MyDistributedTransactionBackend->transactionId.timestamp;
+
+	SpinLockRelease(&MyDistributedTransactionBackend->mutex);
+
+	return currentDistributedTransactionId;
+}
+
+
+/*
+ * GenerateAndSetNextDistributedTransactionId generates a new distributed transaction
+ * id and sets it for the current backend.
+ *
+ * This function should only be called on BeginCoordinatedTransaction(). Any other
+ * callers is very likely to break the distributed transction management.
+ */
+void
+GenerateAndSetNextDistributedTransactionId(void)
+{
+	DistributedTransactionId *distributedTransactionId =
+		GenerateNextDistributedTransactionId();
+
+	SpinLockAcquire(&MyDistributedTransactionBackend->mutex);
+
+	MyDistributedTransactionBackend->databaseId = MyDatabaseId;
+
+	MyDistributedTransactionBackend->transactionId.initiatorNodeIdentifier =
+		distributedTransactionId->initiatorNodeIdentifier;
+	MyDistributedTransactionBackend->transactionId.transactionId =
+		distributedTransactionId->transactionId;
+	MyDistributedTransactionBackend->transactionId.timestamp =
+		distributedTransactionId->timestamp;
+
+	SpinLockRelease(&MyDistributedTransactionBackend->mutex);
+}
+
+
+/*
+ * GenerateNextDistributedTransactionId returns a new distributed transaction id with
+ * the current timestamp and next local transaction id.
+ */
+static DistributedTransactionId *
+GenerateNextDistributedTransactionId(void)
+{
```

Comment:
I think the API split between `GenerateAndSetNextDistributedTransactionId()`, `GenerateNextDistributedTransactionId()`, `GetNextLocalTransactionIdFromShmem()` is a bad idea. You're essentially redoing all the work done in both, adding an extra palloc and a bunch of copying to the mix.  I think we should just collapse all these back into one function.